addons/project/models/project.py

```diff
@@ -0,0 +1,653 @@
+        for task in self.tasks:
+            # preserve task name and stage, normally altered during copy
+            defaults = {'stage_id': task.stage_id.id,
+                        'name': task.name}
+            map_task_id[task.id] =  task.copy(defaults).id
+        project.write({'tasks':[(6,0, map_task_id.values())]})
+
+    @api.multi
+    def copy(self, default=None):
+        default = default or {}
+        if not default.get('name'):
+            default.update(name=_("%s (copy)") % (self.name))
+        project = super(Project, self).copy(default)
+        self.with_context(active_test=False).map_tasks(project)
+        return project
+
+    @api.model
+    def create(self, vals):
+        # Prevent double project creation when 'use_tasks' is checked + alias management
+        create_context = dict(self.env.context, project_creation_in_progress=True,
+                              alias_model_name=vals.get('alias_model', 'project.task'),
+                              alias_parent_model_name=self._name,
+                              mail_create_nosubscribe=True)
+
+        ir_values = self.env['ir.values'].get_default('project.config.settings', 'generate_project_alias')
+        if ir_values:
+            vals['alias_name'] = vals.get('alias_name') or vals.get('name')
+        project = super(Project, self.with_context(create_context)).create(vals)
+        values = {'alias_parent_thread_id': project.id, 'alias_defaults': {'project_id': project.id}}
+        project.alias_id.write(values)
+        return project
+
+    @api.multi
+    def write(self, vals):
+        # if alias_model has been changed, update alias_model_id accordingly
+        if vals.get('alias_model'):
+            models = self.env['ir.model'].search([('model', '=', vals.get('alias_model', 'project.task'))])
+            vals.update(alias_model_id=models.id)
+        res = super(Project, self).write(vals)
+        if 'active' in vals:
+            # archiving/unarchiving a project does it on its tasks, too
+            tasks = self.with_context(active_test=False).mapped('tasks')
+            tasks.write({'active': vals['active']})
+        return res
+
+    @api.multi
+    def unlink(self):
+        alias_ids = []
+        analytic_account_id = set()
+        for project in self:
+            if project.tasks:
+                raise UserError(_('You cannot delete a project containing tasks. You can either delete all the project\'s tasks and then delete the project or simply deactivate the project.'))
+            elif project.alias_id:
+                alias_ids.append(project.alias_id.id)
+            if not project.analytic_account_id.line_ids:
+                analytic_account_id.add(project.analytic_account_id.id)
+        res = super(Project, self).unlink()
+        self.env['mail.alias'].search([('id', 'in', alias_ids)]).unlink()
+        self.env['account.analytic.account'].search([('id', 'in', list(analytic_account_id))]).unlink()
```

Comment:
Why you need search() method here ? Use set operation on recordset(s) and use that recordset(s) to unlink the record(s).